### ISSUE TITLE
delay() and yield() behave differently in test setup than in real life

### ISSUE THREAD

**Issue by Zordid** [Kotlin/kotlinx.coroutines#2605]
For our project, we have developed a special kind of coroutine scope: the InfiniteCoroutineScope.

For the understanding here, I hope it is enough to describe it like this:
- inside this scope, any launched job with the extension fun `launchInfinite(block: ()->Unit)` is monitored never to actually stop. So, multiple "while true" loops are expected to be launched and stay alive infinitely.
- in contrast to a regular coroutineScope, when a single one of those jobs quits working, e.g. ends normally, all other jobs launched via `launchInfinite` will be cancelled and brought to an end.
- so the biggest difference is, that the shortest running job determines the overall run time - and not the longest running job.

This example demonstrates it:
normal coroutineScope behavior:
```
runBlockingTest {
    coroutineScope {
        launch { delay(10_000) }
        launch { delay(5000) }
        launch { delay(4000) }
    }
    assertEquals(10_000, currentTime)
}
```
our infiniteCoroutineScope does this:
```
runBlockingTest {
    infiniteCoroutineScope {
        launchInfinite { delay(10_000) }
        launchInfinite { delay(5000) }
        launchInfinite { delay(4000) }
    }
    assertEquals(4000, currentTime)
}
```
(the shortest running "infinite" job will kill all other's)

In *real* life this of course also works very fine when instead of using delay, yield is called in an endless loop to allow cooperative coroutine behaviour: 
```
infiniteCoroutineScope {
    launchInfinite { while (true) yield() }
    launchInfinite { delay(5000) }
    launchInfinite { delay(4000) }
}
// time spent in this block above will also be 4000ms!
```

*BUT*: this in a test does not work at all. It runs forever and does not terminate. So, as far as I understand: there must be a conceptional difference in runBlockingTest in such that yield and delay do not automatically do the same thing in terms of eager execution and virtual time. But I do not know what that might be...

All I can see is that even by just replacing
`while (true) yield()` with `while (true) delay(1)` makes it all work fine, while `delay(0)` breaks it, too!

And that is my problem: I always took `yield()` as a "non existent delay" to allow coroutines to properly work together and used this instead of a not needed delay of an arbitrarily small amount of time.

Could it be, that for coroutines-test, yield tries to auto-advance virtual time by 0 and as such, we are stuck in my endless loop and never actually get forward in time?

EDIT:
I proved my assumption above by outputting the current virtual time - and was spot on. In an endless loop with yield, the `currentTime` stays at 0 forever. Replacing yield with delay and a value >0 does the trick. Time passes, and everything works just fine!

**Comment by dkhalanskyjb** [comment]
Looks like the same behavior can be observed on `Dispachers.Unconfined`, right? Though `Dispatchers.Unconfined` and `TestCoroutineDispatcher` behave this way for different reasons.

**Comment by Orgmir** [comment]
@Zordid Do you end up writing tests for the while loop case that don't involve the `delay(1)`?

I find myself in the a similar situation, trying to write some tests with a while loop, and following this issue I ended up using `delay(1)` instead of `yield()` to make the tests work.

**Comment by dkhalanskyjb** [comment]
In the end, I couldn't reproduce the issue. For example, the following test passes for me:
```kotlin
    @Test
    fun testWhileYield() = runBlockingTest {
        var x = 0
        val job1 = launch {
            while (true) {
                yield()
                ++x
            }
        }
        var y = 0
        val job2 = launch {
            while (true) {
                yield()
                ++y
            }
        }
        val job3 = launch {
            while(x < 100 && y < 100) {
                yield()
            }
            job1.cancelAndJoin()
            job2.cancelAndJoin()
        }
        job3.join()
        assertFalse(job1.isActive)
        assertFalse(job2.isActive)
        assertFalse(job3.isActive)
    }
```
It does look like the coroutines cooperate just fine: `job1` increments `x`, `job2` increments `y`, and `job3` waits for them to do this, and then cancels everything.

That said, the issue is likely fixed now: since `1.6.0-RC`, a new API for the test framework is available, which fixes a lot of problems with the confusing order of coroutine execution.

I'll close the issue, but if you still encounter it with the new API, please feel free to reopen it.

**Comment by Orgmir** [comment]
@dkhalanskyjb thank you for your example! I think I figured it out and got my tests working.

I should have been more specific, I was using `1.6.0-RC` and `runTest` instead of `runBlockingTest`. The bit that made it work was canceling the job inside another coroutine.

Here is an example of what I'm doing:

```kotlin
@Test
    fun testThatItYields() = runTest {
        var didFetch = false
        val job = launch {
            while(true) {
                didFetch = true 
                yield()
            }
        }

        val cancelJob = launch { 
            job.cancelAndJoin()
        }
        
        cancelJob.join()
        assertTrue(didFetch)
    }
```

Before I was just calling `job.cancel()` after the job was created, so I think my issue was the job was dead before it got a chance to run.
